### ISSUE TITLE
WebDriver BiDi `browsingContext` events tests are racy

### ISSUE THREAD

**Issue by sadym-chromium** [web-platform-tests/wpt#35846]
The way tests are relying on the loading events e.g. in line [`on_entry = wait_for_event(CONTEXT_LOAD_EVENT)`](https://github.com/web-platform-tests/wpt/blob/09f51f2cf4a939df018339d83d1c59ce678f51fa/webdriver/tests/bidi/browsing_context/load/load.py#L42) is racy. There are no guarantee at that point, that the initial new tab is loaded. And, therefore, the assertion [`    assert_navigation_info(event, new_tab["context"], url)`](https://github.com/web-platform-tests/wpt/blob/09f51f2cf4a939df018339d83d1c59ce678f51fa/webdriver/tests/bidi/browsing_context/load/load.py#L47) will fail. That is what actually happens to the ChromeDriverBiDi implementation.

We should update tests and wait for the new tab to be loaded in [` new_tab` fixture](https://github.com/web-platform-tests/wpt/blob/09f51f2cf4a939df018339d83d1c59ce678f51fa/webdriver/tests/support/fixtures_bidi.py#L12).

Another option is to update the protocol behaviour, but it seems a bit odd.

**Comment by sadym-chromium** [comment]
Attempt to solve the issue by adding an additional navigation into [`new_tab` fixture](https://github.com/web-platform-tests/wpt/blob/09f51f2cf4a939df018339d83d1c59ce678f51fa/webdriver/tests/support/fixtures_bidi.py#L12): https://github.com/web-platform-tests/wpt/pull/35848

**Comment by whimboo** [comment]
Based on several existing WebDriver BiDi tests around `browsingContext.create` and `browsingContext.contextCreated` all the assertions that I can see in the test files reference `about:blank` as URL. Given that page load events are emitted after the `browsingContext.contextCreated` event I cannot see why this should be racy. Means the initial `about:blank` should always be loaded.

CC'ing @jgraham for further feedback. Maybe I'm missing something here...

**Comment by jgraham** [comment]
Hmm, so about:blank is tricky, and gecko are trying to change their behaviour so that the load is synchronous like Blink and WebKit. Given the current HTML spec, it's not clear if you can expect to observe a load event for the initial about:blank, so I think it makes some amount of sense to not have our tests depend on this tricky area of interop. I think for now it would be OK to just ignore any load event if the URL is `about:blank`, which I think would avoid the possible race without adding in yet another load.

**Comment by sadym-chromium** [comment]
> I think for now it would be OK to just ignore any load event if the URL is `about:blank`, which I think would avoid the possible race without adding in yet another load.

@jgraham  Do you mean to filter out the load events for `about:blank` in the WPT tests, or on the protocol level?

**Comment by whimboo** [comment]
> for now it would be OK to just ignore any load event if the URL is `about:blank`, which I think would avoid the possible race without adding in yet another load.

I assume this was meant to only apply for newly opened browsing contexts but not following navigation commands? The latter should still emit navigation events.

IMHO it would be great to have some content in the BiDi spec which clarifies this.

**Comment by sadym-chromium** [comment]
> Given that page load events are emitted after the `browsingContext.contextCreated` event I cannot see why this should be racy. Means the initial `about:blank` should always be loaded.

@whimboo I see the step [**"23. Completely finish loading document."**](https://html.spec.whatwg.org/multipage/browsers.html#creating-browsing-contexts) of the [**"create a new browsing context"**](https://html.spec.whatwg.org/multipage/browsers.html#creating-a-new-browsing-context) procedure. But BiDi spec uses the [**"window open steps"**](https://html.spec.whatwg.org/multipage/window-object.html#window-open-steps), which in order doesn't wait for the page (`about:blank`) to be loaded.

It looks like adding something like **_"Wait for the browsing context to be loaded"_** in the [**"browsingContext.create"**](https://w3c.github.io/webdriver-bidi/#command-browsingContext-create) before the last step **"Return success with data body"** should solve the issue, as all the load events would be emitted before the command [`await bidi_session.browsing_context.create(type_hint='tab')`](https://github.com/web-platform-tests/wpt/blob/7708eea45e7015c3548d0f63c6d5ce5008bf79be/webdriver/tests/support/fixtures_bidi.py#L12) is done.

**Comment by jgraham** [comment]
> It looks like adding something like "Wait for the browsing context to be loaded" in the ["browsingContext.create"](https://w3c.github.io/webdriver-bidi/#command-browsingContext-create) before the last step "Return success with data body" should solve the issue, as all the load events would be emitted before the command [await bidi_session.browsing_context.create(type_hint='tab')](https://github.com/web-platform-tests/wpt/blob/7708eea45e7015c3548d0f63c6d5ce5008bf79be/webdriver/tests/support/fixtures_bidi.py#L12) is done.

This seems like the wrong solution. I think we want the `contextCreated` event to always precede any other event for that context, so that you can rely on the invariant that if you get an event with a `context` parameter you will always have got a `contextCreated` event for that context (assuming you're subscribed to those events).

> IMHO it would be great to have some content in the BiDi spec which clarifies this.

I don't want to poke too hard at the spec in this area; https://github.com/whatwg/html/pull/6315 is a long-running effort to make the HTML spec match implementations, and once that lands it makes more sense to ensure that BiDi is correct.

What I think we could reasonably do is document the invariant above about `contextCreated` firing before any other events relating to that context.

**Comment by sadym-chromium** [comment]
I updated the PR to wait for the new tab to be loaded. It should fix the problem: https://github.com/web-platform-tests/wpt/pull/35848

**Comment by sadym-chromium** [comment]
> Based on several existing WebDriver BiDi tests around `browsingContext.create` and `browsingContext.contextCreated` all the assertions that I can see in the test files reference `about:blank` as URL. Given that page load events are emitted after the `browsingContext.contextCreated` event I cannot see why this should be racy. Means the initial `about:blank` should always be loaded.

The tests does the following:
1. Create new tab and wait the command to be finished.
2. Add a listener for the `browsingContext.load` events: [`on_entry = wait_for_event(CONTEXT_LOAD_EVENT)`](https://github.com/web-platform-tests/wpt/blob/9c976530d618f2e9810d5284b607dfb5006285e7/webdriver/tests/bidi/browsing_context/load/load.py#L42).
3. Navigate the new tab: [`await bidi_session.browsing_context.navigate(context=new_tab["context"], url=url)`](https://github.com/web-platform-tests/wpt/blob/9c976530d618f2e9810d5284b607dfb5006285e7/webdriver/tests/bidi/browsing_context/load/load.py#L44)
4. Wait for the `browsingContext.load` event: [`event = await on_entry`](https://github.com/web-platform-tests/wpt/blob/9c976530d618f2e9810d5284b607dfb5006285e7/webdriver/tests/bidi/browsing_context/load/load.py#L45).
5. Assert the event is the correct one: [`assert_navigation_info(event, new_tab["context"], url)`](https://github.com/web-platform-tests/wpt/blob/9c976530d618f2e9810d5284b607dfb5006285e7/webdriver/tests/bidi/browsing_context/load/load.py#L47).

The problem is in the fact the `browsingContext.load` event for the initial new tab navigation to the `about:blank` can be raised after step **2. Add a listener for the `browsingContext.load` events**.

**Comment by juliandescottes** [comment]
>  I think we want the contextCreated event to always precede any other event for that context, so that you can rely on the invariant that if you get an event with a context parameter you will always have got a contextCreated event for that context (assuming you're subscribed to those events).

We could keep the invariant on the events, guarantee that contextCreated is emitted before any other event, but still block resolving `browsingContext.create` until we got the corresponding `contextCreated` event. This doesn't have to impact event ordering? Would that work?

**Comment by sadym-chromium** [comment]
Closing in favour of https://github.com/w3c/webdriver-bidi/issues/288
